File: rostam-vpn-apple-develop/Sources/WireGuardApp/Tunnel/ConfigBuilder.swift
// SPDX-License-Identifier: MIT
// Copyright © 2018-2019 WireGuard LLC. All Rights Reserved.

import Foundation
import Amplitude

class ConfigBuilder {
    private static let apiDataKey = "rostamApiData"
    private static let apiExpiryKey = "rostamApiExpiry"
    private static var path = ""
//    private static var host = ""
    private static var domainFront = ""
    private static var apiUrl = "https://docs.google.com/feeds/download/documents/export/Export?id=1eU2garmd4ZKLCEGDoqpQ8I2-C7llC8u1chJ3uo1OPt8&exportFormat=html"

    private static var expiry = 0

    static func build(privateKey: PrivateKey, publicKey: PublicKey, region: String?, completionHandler: @escaping ((String?) -> Void)) {
        getApiData { success, error in
            if success {
                getProfileData(publicKey: publicKey, region: region) { apiResponse, errorMessage in
                    if let apiResponse = apiResponse {
                        let endpoint: String = apiResponse.endpoints[0]
                        let wgQuickConfig = createWqQuickConfig(privateKey: privateKey.base64Key, publicKey: apiResponse.publicKey, address: apiResponse.address, endpoint: endpoint)

                        Amplitude.instance().logEvent("Call to profile API succeeded")
                        completionHandler(wgQuickConfig)
                    } else {
                        if let errorMessage = errorMessage {
                            if !errorMessage.isEmpty {
                                let eventProperties = ["error": errorMessage as Any]
                                Amplitude.instance().logEvent("Call to profile API error", withEventProperties: eventProperties)
                            }
                        }

                        Amplitude.instance().logEvent("Call to profile API failed")
                        completionHandler(nil)
                    }
                }
            } else {
                if let error = error, !error.isEmpty {
                    let eventProperties = ["error": error as Any]
                    Amplitude.instance().logEvent("Failed to obtain API data JSON", withEventProperties: eventProperties)
                } else {
                    Amplitude.instance().logEvent("Failed to obtain API data JSON")
                }
                completionHandler(nil)
            }
        }
    }

    static func parse(privateKey: PrivateKey, pubkey: PublicKey, fileContents: String) -> String? {
        let jsonString = "\(fileContents.replacingOccurrences(of: "'", with: "\""))"
        let data = jsonString.data(using: .utf8)!
        guard let json = try? JSONSerialization.jsonObject(with: data, options: .allowFragments) as? [String: Any] else { return nil }
        debugPrint(json)

        if let address = json["address"] as? String,
           let endpoints = json["endpoint"] as? [String],
           let clientPublicKey = json["client_pubkey"] as? String,
           let serverPublicKey = json["pubkey"] as? String {
            if clientPublicKey == pubkey.base64Key {
                EndpointManager.storeEndpoints(endpoints: endpoints, region: nil)

                let endpoint: String = endpoints[0]
                let wgQuickConfig = createWqQuickConfig(privateKey: privateKey.base64Key, publicKey: serverPublicKey, address: address, endpoint: endpoint)

                return wgQuickConfig
            }
        }

        return nil
    }

    private static func getProfileData(publicKey: PublicKey, region: String?, completionHandler: @escaping (ApiResponse?, String?) -> Void) {
        let profileApiUrl = "https://" + domainFront + path

        var request = URLRequest(url: URL(string: profileApiUrl)!)
        request.httpMethod = "POST"
        request.addValue("application/json", forHTTPHeaderField: "Content-Type")
        request.addValue("application/json", forHTTPHeaderField: "Accept")
//        request.addValue(host, forHTTPHeaderField: "Host")
        request.timeoutInterval = 5
        request.cachePolicy = .reloadIgnoringLocalAndRemoteCacheData
        var parameterDictionary = ["pubkey": publicKey.base64Key]
        if let region = region {
            parameterDictionary["region"] = region
        }
        guard let httpBody = try? JSONSerialization.data(withJSONObject: parameterDictionary, options: []) else {
            completionHandler(nil, "JSON serialization error")
            return
        }
        request.httpBody = httpBody

        let urlSession = URLSession.shared
        let task = urlSession.dataTask(with: request) { data, response, error -> Void in
            var errorMessage = ""
            if let error = error {
                debugPrint(error)
                errorMessage = error.localizedDescription
            } else if let data = data,
                let response = response as? HTTPURLResponse,
                response.statusCode == 200 {
                debugPrint(response)
                do {
                    guard let json = try JSONSerialization.jsonObject(with: data) as? [String: Any] else { return }
                    debugPrint(json)

                    if let status = json["status"] as? String,
                        status == "ok",
                        let message = json["message"] as? String,
                        let address = json["address"] as? String,
                        let endpoints = json["endpoint"] as? [String], let serverPublicKey = json["pubkey"] as? String {

                        EndpointManager.storeEndpoints(endpoints: endpoints, region: region)
                        let apiResponse = ApiResponse(status: status, message: message, address: address, endpoints: endpoints, publicKey: serverPublicKey)
                        completionHandler(apiResponse, nil)
                        return
                    } else if let message = json["message"] as? String {
                        debugPrint(message)
                        errorMessage = message
                    }
                } catch let error {
                    debugPrint(error)
                    errorMessage = error.localizedDescription
                }
            } else {
                errorMessage = "Invalid URL"
            }

            completionHandler(nil, errorMessage)
        }

        task.resume()
    }

    private static func getApiData(completionHandler: @escaping (Bool, String?) -> Void) {
        if !isExpired() {
            if let jsonString = UserDefaults.standard.value(forKey: apiDataKey) as? String {
                do {
                    try parseApiDataJson(jsonString: jsonString)

                    completionHandler(true, nil)
                    return
                } catch let error {
                    debugPrint(error)
                }
            }
        }

        var request = URLRequest(url: URL(string: apiUrl)!)
        request.httpMethod = "GET"
        request.addValue("application/json", forHTTPHeaderField: "Content-Type")
        request.addValue("application/json", forHTTPHeaderField: "Accept")
        request.timeoutInterval = 5
        request.cachePolicy = .reloadIgnoringLocalAndRemoteCacheData

        let urlSession = URLSession.shared
        let task = urlSession.dataTask(with: request) { _, response, error -> Void in
            var errorMessage = ""
            if let error = error {
                debugPrint(error)
                errorMessage = error.localizedDescription
            } else if let response = response as? HTTPURLResponse,
                  response.statusCode == 200,
                  let filename = response.suggestedFilename {

                let base64String: String = filename.components(separatedBy: ".")[0]
                if let data = Data(base64Encoded: base64String),
                   let jsonString = String(data: data, encoding: .utf8) {

                    do {
                        try parseApiDataJson(jsonString: jsonString)
                        storeExpiry(expiry: expiry)
                        storeApiData(jsonString: jsonString)

                        completionHandler(true, nil)
                        return
                    } catch let error {
                        debugPrint(error)
                        errorMessage = error.localizedDescription
                    }
                }

            }

            completionHandler(false, errorMessage)
        }

        task.resume()
    }

    private static func createWqQuickConfig(privateKey: String, publicKey: String, address: String, endpoint: String) -> String {
        var wgQuickConfig = "[Interface]\n"
        wgQuickConfig.append("PrivateKey = \(privateKey)\n")
        wgQuickConfig.append("Address = \(address)\n")
        wgQuickConfig.append("DNS = 8.8.8.8\n")
        wgQuickConfig.append("\n[Peer]\n")
        wgQuickConfig.append("PublicKey = \(publicKey)\n")
        wgQuickConfig.append("AllowedIPs = 0.0.0.0/0\n")
        wgQuickConfig.append("Endpoint = \(endpoint)\n")
        wgQuickConfig.append("PersistentKeepalive = 25")

        return wgQuickConfig
    }

    private static func storeApiData(jsonString: String) {
        UserDefaults.standard.set(jsonString, forKey: apiDataKey)
    }

    private static func storeExpiry(expiry: Int) {
        let calendar = Calendar.current
        let expiryDate = calendar.date(byAdding: .second, value: expiry, to: Date())
        UserDefaults.standard.set(expiryDate, forKey: apiExpiryKey)
    }

    private static func isExpired() -> Bool {
        if let expiryDate = UserDefaults.standard.value(forKey: apiExpiryKey) as? Date {
            let isExpired = Date() > expiryDate

            return isExpired
        }

        return true
    }

    private static func parseApiDataJson(jsonString: String) throws {
        if let jsonData = jsonString.data(using: .utf8) {
            if let json = try JSONSerialization.jsonObject(with: jsonData) as? [String: Any],
                let path = json["path"] as? String,
//                let host = json["host"] as? String,
                let domainFront = json["domain_front"] as? String,
                let apiUrl = json["apiURL"] as? String,
                let expiry = json["expiry"] as? Int {

                self.path = path
//                self.host = host
                self.domainFront = domainFront
                self.apiUrl = apiUrl
                self.expiry = expiry

                Amplitude.instance().logEvent("Successfully obtained API data JSON", withEventProperties: json)
                return
            }
        }

        Amplitude.instance().logEvent("Failed to parse API data JSON")
    }
}

struct ApiResponse {
    private(set) var status: String
    private(set) var message: String
    private(set) var address: String
    private(set) var endpoints: [String]
    private(set) var publicKey: String

    init(status: String, message: String, address: String, endpoints: [String], publicKey: String) {
        self.status = status
        self.message = message
        self.address = address
        self.endpoints = endpoints
        self.publicKey = publicKey
    }
}

File: rostam-vpn-develop/android/ui/src/main/java/com/rostamvpn/android/rostamProfile/ConfigBuilder.java
package com.rostamvpn.android.rostamProfile;

import android.content.Context;
import android.content.SharedPreferences;
import android.util.Base64;
import android.util.Log;

import com.amplitude.api.Amplitude;

import com.android.volley.Request;
import com.android.volley.RequestQueue;
import com.android.volley.toolbox.JsonObjectRequest;

import com.rostamvpn.android.Application;
import com.rostamvpn.android.util.FilenameRequest;
import com.rostamvpn.config.BadConfigException;
import com.rostamvpn.config.Config;
import com.rostamvpn.crypto.KeyPair;
import com.rostamvpn.util.NonNullForAll;

import java.io.BufferedReader;
import java.io.ByteArrayInputStream;
import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.net.InetAddress;
import java.net.URL;
import java.net.UnknownHostException;
import java.nio.charset.StandardCharsets;
import java.security.InvalidKeyException;
import java.util.Date;
import java.util.HashMap;
import java.util.Map;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

@NonNullForAll
public class ConfigBuilder {
    private static final String TAG = "RostamVPN/" + ConfigBuilder.class.getSimpleName();
    private static final String API_DATA_TAG = "RostamVPNApiData";
    private static final String PROFILE_API_TAG = "RostamVPNProfileApi";
    private static final String API_DATA_KEY = "rostam_api_data";
    private static final String API_EXPIRY_KEY = "rostam_api_expiry";
    private static final String DEFAULT_API_URL = "https://docs.google.com/feeds/download/documents/export/Export?id=1eU2garmd4ZKLCEGDoqpQ8I2-C7llC8u1chJ3uo1OPt8&amp;exportFormat=html";
    private static String path;
    private static String host;
    private static String domainFront;
    private static String apiUrl;
    private static int expiry = 0;

    public static void build(final Context context, final KeyPair keyPair, final String region, ConfigBuilderCallback callback) {
        getApiData(context, new ApiDataCallback() {
            @Override
            public void onSuccess() {
                getProfileData(keyPair, region, new ProfileApiCallback() {
                    @Override
                    public void onSuccess(final String address, final String[] endpoints, final String serverPublicKey) {
                        final String endpoint = endpoints[0];
                        final String configText = createConfigText(keyPair.getPrivateKey().toBase64(), serverPublicKey, address, endpoint);

                        Amplitude.getInstance().logEvent("Call to profile API succeeded");

                        callback.onSuccess(configText);
                    }

                    @Override
                    public void onFail(final String errorMessage) {
                        try {
                            JSONObject eventProperties = new JSONObject();
                            eventProperties.put("error", errorMessage);
                            Amplitude.getInstance().logEvent("Call to profile API error", eventProperties);
                        }
                        catch (JSONException ex) {
                            Log.e(TAG, ex.getMessage());
                        }

                        Amplitude.getInstance().logEvent("Call to profile API failed");

                        callback.onFail();
                    }
                });
            }

            @Override
            public void onFail(final String errorMessage) {
                try {
                    JSONObject eventProperties = new JSONObject();
                    eventProperties.put("error", errorMessage);
                    Amplitude.getInstance().logEvent("Failed to obtain API data JSON", eventProperties);
                }
                catch (JSONException ex) {
                    Log.e(TAG, ex.getMessage());
                }

                callback.onFail();
            }
        });
    }

    public static Config parse(final KeyPair keyPair, final InputStream stream) throws IOException, JSONException, BadConfigException, InvalidKeyException {
        final StringBuilder builder = new StringBuilder();
        final BufferedReader reader = new BufferedReader(new InputStreamReader(stream));

        String line;
        while ((line = reader.readLine()) != null) {
            builder.append(line + System.lineSeparator());
        }

        final JSONObject json = new JSONObject(builder.toString());

        final String address = json.getString( "address");
        final String[] endpoints = getStringArray(json.getJSONArray( "endpoint"));
        final String clientPublicKey = json.getString("client_pubkey");
        final String serverPublicKey = json.getString( "pubkey");

        final String pubkey = keyPair.getPublicKey().toBase64();
        if(!clientPublicKey.equals(pubkey)) {
            throw new InvalidKeyException("Invalid public key!");
        }

        EndpointManager.storeEndpoints(endpoints, null);

        final String endpoint = endpoints[0];
        final String configText = createConfigText(keyPair.getPrivateKey().toBase64(), serverPublicKey, address, endpoint);
        final Config config = Config.parse(new ByteArrayInputStream(configText.getBytes(StandardCharsets.UTF_8)));

        return config;
    }

    private static void getProfileData(final KeyPair keyPair, final String region, ProfileApiCallback callback) {
        final String profileApiUrl = "https://" + domainFront + path;

        // This is a hack for the UnknownHostException issue
        // https://stackoverflow.com/questions/3293659/android-java-net-unknownhostexception-host-is-unresolved-strategy-question
        Application.getAsyncWorker().runAsync(() -> {
            try {
                URL url = new URL(profileApiUrl);
                InetAddress inetAddress = InetAddress.getByName(url.getHost());
            } catch (UnknownHostException e) {
                Log.e(TAG, e.getMessage());
            }
        });

        // Get the RequestQueue.
        RequestQueue queue = Application.getRequestQueue();
        queue.getCache().clear();

        // Call the API to get the endpoint and address...
        Map<String, String> params = new HashMap<>();
        params.put("pubkey", keyPair.getPublicKey().toBase64());
        if(region != null) {
            params.put("region", region);
        }
        JSONObject parameters = new JSONObject(params);
        JsonObjectRequest jsonObjectRequest = new JsonObjectRequest
                (Request.Method.POST, profileApiUrl, parameters, response -> {
                    Log.d(TAG, "Profile API Response: " + response);

                    String status;
                    String message = "";
                    String address = "";
                    String[] endpoints = null;
                    String serverPublicKey = "";
                    try {
                        status = response.getString("status");
                        message = response.getString( "message");
                        address = response.getString( "address");
                        endpoints = getStringArray(response.getJSONArray( "endpoint"));
                        serverPublicKey = response.getString( "pubkey");
                    }
                    catch (JSONException e) {
                        Log.e(TAG, e.getMessage());
                        status = "fail";
                    }

                    if(status.equals("ok")) {
                        EndpointManager.storeEndpoints(endpoints, region);
                        callback.onSuccess(address, endpoints, serverPublicKey);
                    }
                    else {
                        Log.d(TAG, message);
                        callback.onFail(message);
                    }
                }, error -> {
                    Log.e(TAG, error.getMessage(), error);
                    final String message = error.getMessage() != null ? error.getMessage() : error.toString();
                    callback.onFail(message);
                }) {
            @Override
            public Map<String, String> getHeaders() {
                Map<String, String>  headers = new HashMap<>();
                headers.put("Host", host);

                return headers;
            }
        };

        // Add the request to the RequestQueue.
        jsonObjectRequest.setTag(PROFILE_API_TAG);
        jsonObjectRequest.setRetryPolicy(Application.getDefaultRetryPolicy());
        jsonObjectRequest.setShouldCache(false);
        queue.add(jsonObjectRequest);
    }

    private static void getApiData(final Context context, ApiDataCallback callback) {
        if(!isExpired()) {
            final String jsonString = Application.getSharedPreferences().getString(API_DATA_KEY, null);
            if(jsonString != null) {
                if(parseApiDataJson(jsonString) == true) {
                    callback.onSuccess();
                    return;
                }
            }
        }

        final String apiDataUrl = apiUrl == null ? DEFAULT_API_URL : apiUrl;
        RequestQueue queue = Application.getRequestQueue();
        queue.getCache().clear();

        FilenameRequest filenameRequest = new FilenameRequest
            (apiDataUrl, response -> {
                Log.d(TAG, "API data Response: " + response);

                byte[] decodedBytes = Base64.decode(response, Base64.DEFAULT);
                String jsonString = new String(decodedBytes);
                Log.d(TAG, "API data JSON: " + jsonString);

                if (parseApiDataJson(jsonString) == true) {
                    storeExpiry(expiry);
                    storeApiData(jsonString);
                    callback.onSuccess();
                } else {
                    callback.onFail("Failed to parse API data JSON");
                }
            }, error -> {
                final String message = error.getMessage() != null ? error.getMessage() : error.toString();
                Log.e(TAG, message);
                callback.onFail(message);
            });

        // Add the request to the RequestQueue.
        filenameRequest.setTag(API_DATA_TAG);
        filenameRequest.setRetryPolicy(Application.getDefaultRetryPolicy());
        filenameRequest.setShouldCache(false);
        queue.add(filenameRequest);
    }
    
    private static String createConfigText(final String privateKey, final String publicKey, final String address, final String endpoint) {
        final StringBuilder builder = new StringBuilder();
        builder.append("[Interface]");
        builder.append("\nPrivateKey = " + privateKey);
        builder.append("\nAddress = " + address);
        builder.append("\nDNS = 8.8.8.8");
        builder.append("\n\n[Peer]");
        builder.append("\nPublicKey = " + publicKey);
        builder.append("\nAllowedIPs = 0.0.0.0/0");
        builder.append("\nEndpoint = " + endpoint);
        builder.append("\nPersistentKeepalive = 25");

        return builder.toString();
    }

    private static String[] getStringArray(JSONArray jsonArray) {
        String[] stringArray = null;
        if (jsonArray != null) {
            int length = jsonArray.length();
            stringArray = new String[length];
            for (int i = 0; i < length; i++) {
                stringArray[i] = jsonArray.optString(i);
            }
        }

        return stringArray;
    }

    private static void storeApiData(final String json) {
        SharedPreferences sharedPreferences = Application.getSharedPreferences();
        SharedPreferences.Editor editor = sharedPreferences.edit();
        editor.putString(API_DATA_KEY, json);
        editor.apply();
    }

    private static void storeExpiry(final int expiry) {
        SharedPreferences sharedPreferences = Application.getSharedPreferences();
        SharedPreferences.Editor editor = sharedPreferences.edit();
        Date now = new Date();
        long expiryLong = now.getTime() + expiry * 1000;
        editor.putLong(API_EXPIRY_KEY, expiryLong);
        editor.apply();
    }

    private static boolean isExpired() {
        final long expiryLong = Application.getSharedPreferences().getLong(API_EXPIRY_KEY, 0);
        if(expiryLong == 0) return true;

        Date now = new Date();
        return now.getTime() > expiryLong;
    }

    private static boolean parseApiDataJson(final String jsonString) {
        final JSONObject json;
        try {
            json = new JSONObject(jsonString);

            path = json.getString( "path");
            host = json.getString( "host");
            domainFront = json.getString( "domain_front");
            apiUrl = json.getString( "apiURL");
            expiry = json.getInt( "expiry");

            Amplitude.getInstance().logEvent("Successfully obtained API data JSON", json);

            return true;
        } catch (JSONException e) {
            Log.e(TAG, e.getMessage());
            return false;
        }
    }
}